lib/oli_web/controllers/delivery_controller.ex

```diff
@@ -400,41 +400,30 @@ defmodule OliWeb.DeliveryController do
   end
 
   def show_enroll(conn, params) do
-    case Sections.available?(conn.assigns.section) do
-      {:available, section} ->
-        # redirect to course index if user is already signed in and enrolled
-        with {:ok, user} <- current_or_guest_user(conn, section.requires_enrollment),
-             true <- Sections.is_enrolled?(user.id, section.slug) do
-          redirect(conn,
-            to: ~p"/sections/#{section.slug}"
-          )
-        else
-          {:redirect, nil} ->
-            # guest user cant access courses that require enrollment
-            login_params = [
-              section: section.slug,
-              from_invitation_link?: params["from_invitation_link?"] || false
-            ]
-
-            redirect_path =
-              ~p"/session/new?#{login_params}"
-
-            conn
-            |> redirect(to: redirect_path)
-
-          _ ->
-            section = Oli.Repo.preload(section, [:base_project])
+    section = conn.assigns.section
+    from_invitation_link? = params["from_invitation_link?"] || false
+
+    with {:available, section} <- Sections.available?(section),
+         {:ok, user} <- current_or_guest_user(conn, section.requires_enrollment),
+         {:enrolled?, false} <- {:enrolled?, Sections.is_enrolled?(user.id, section.slug)} do
+      render(conn, "enroll.html",
+        section: Oli.Repo.preload(section, [:base_project]),
+        from_invitation_link?: from_invitation_link?,
+        auto_enroll_as_guest: params["auto_enroll_as_guest"] || false
+      )
+    else
+      {:unavailable, reason} ->
+        render_section_unavailable(conn, reason)
 
-            render(conn, "enroll.html",
-              section: section,
-              from_invitation_link?: params["from_invitation_link?"] || false,
-              auto_enroll_as_guest: params["auto_enroll_as_guest"] || false
-            )
-        end
+      # redirect to course index if user is already signed in and enrolled
+      {:enrolled?, true} ->
```

Comment:
Thanks for the review!
I used atoms as the first element of the tuple to more clearly indicate which case is failing. Since the happy path here is that the user is not enrolled, it might be a bit challenging to understand if we only pattern match the else case with a single `true -> ...`. Additionally, I used atoms with a trailing question mark to follow a similar pattern, like in [this example from Ecto](https://github.com/elixir-ecto/ecto/blob/v3.12.1/lib/ecto/changeset.ex#L369).

However, if you would prefer a different approach or if you think it would improve clarity, I'm happy to remove or adjust them.

Reply:
Ah interesting, I see now that it is paired with a value of true/false and I suppose given this pattern is used in Ecto, this is probably okay